enclave_build/src/extract.rs

```diff
@@ -0,0 +1,415 @@
+
+    // Try to parse the config JSON for the 'Env' field
+    match json_object
+        .get("container_config")
+            .ok_or_else(|| ExtractError::EnvCmdError("'container_config' field is missing in the config JSON.".to_string()))?
+        .get("Env")
+            .ok_or_else(|| ExtractError::EnvCmdError("'Env' field is missing in the configuration JSON.".to_string()))?
+        // Try to extract the array of ENV expressions
+        .as_array() {
+            None => Err(ExtractError::EnvCmdError("Failed to extract ENV expressions from image.".to_string())),
+            Some(env_array) => {
+                let env_strings: Vec<String> = env_array.iter().map(|json_value| {
+                    let mut string = json_value.to_string();
+                    // Remove the quotes from the beginning and the end of the expression
+                    string.pop();
+                    string.remove(0);
+                    string
+                }).collect();
+                Ok(env_strings)
+            }
+    }
+}
+
+/// Extract the CMD expressions from an image
+pub fn extract_cmd_expressions(image_data: &ImageData) -> Result<Vec<String>> {
+    let config_string = String::from_utf8(image_data.config.data.clone())
+        .map_err(|err| ExtractError::EnvCmdError(format!(
+            "Failed to extract 'CMD' expressions: {:?}", err)))?;
+
+    let json_object: Value = serde_json::from_str(config_string.as_str()).map_err(|err|
+        ExtractError::EnvCmdError(format!("Failed to extract CMD expressions from an image: {:?}", err)))?;
+
+    // Try to parse the config JSON for the 'Cmd' field
+    match json_object
+        .get("container_config")
+            .ok_or_else(|| ExtractError::EnvCmdError("'container_config' field is missing in the config JSON.".to_string()))?
+        .get("Cmd")
+            .ok_or_else(|| ExtractError::EnvCmdError("'Cmd' field is missing in the configuration JSON.".to_string()))?
+        // Try to extract the array of CMD expressions
+        .as_array() {
+            None => Err(ExtractError::EnvCmdError("Failed to extract CMD expressions from image.".to_string())),
+            Some(cmd_array) => {
+                let cmd_strings: Vec<String> = cmd_array.iter().map(|json_value| {
+                    let mut string = json_value.to_string();
+                    // Remove the quotes from the beginning and the end of the expression
+                    string.pop();
+                    string.remove(0);
+                    string
+                }).collect();
+                Ok(cmd_strings)
+            }
+    }
+}
+
+/// Extract the image hash (digest) from an image
+pub fn extract_image_hash(image_data: &ImageData) -> Result<String> {
+    // Extract the config JSON from the image
+    let config_json = extract_config_json(image_data)
+        .map_err(|err| ExtractError::ImageHashError(format!("{:?}", err)))?;
```

Comment:
This could have easily been:
```
 .map_err(|err| ExtractError::ImageHashError(err))?
```

Same for ConfigError(String) EnvCmdError(String)